nilearn/connectome/tests/test_connectivity_matrices.py

```diff
@@ -425,208 +587,196 @@ def test_prec_to_partial():
-                            gmean_sqrt
-                        ),
-                        covs[k],
-                    )
-                elif kind == "precision":
-                    assert is_spd(cov_new, decimal=7)
-                    assert_array_almost_equal(
-                        cov_new.dot(covs[k]), np.eye(n_features)
-                    )
-                elif kind == "correlation":
-                    assert is_spd(cov_new, decimal=7)
-                    d = np.sqrt(np.diag(np.diag(covs[k])))
-                    if cov_estimator == EmpiricalCovariance():
-                        assert_array_almost_equal(
-                            d.dot(cov_new).dot(d), covs[k]
-                        )
-                    assert_array_almost_equal(
-                        np.diag(cov_new), np.ones(n_features)
-                    )
-                elif kind == "partial correlation":
-                    prec = linalg.inv(covs[k])
-                    d = np.sqrt(np.diag(np.diag(prec)))
-                    assert_array_almost_equal(
-                        d.dot(cov_new).dot(d),
-                        -prec + 2 * np.diag(np.diag(prec)),
-                    )
-
-    # Check the mean_
-    for kind in kinds:
-        conn_measure = ConnectivityMeasure(kind=kind)
-        conn_measure.fit_transform(signals)
-        assert (conn_measure.mean_).shape == (n_features, n_features)
-        if kind != "tangent":
-            assert_array_almost_equal(
-                conn_measure.mean_,
-                np.mean(conn_measure.transform(signals), axis=0),
-            )
+    input_covs = copy.copy(covs)
+    conn_measure = ConnectivityMeasure(kind=kind, cov_estimator=cov_estimator)
+    connectivities = conn_measure.fit_transform(signals)
+
+    # Generic
+    assert isinstance(connectivities, np.ndarray)
+    assert len(connectivities) == len(covs)
+
+    for k, _ in enumerate(connectivities):
+        assert_array_equal(input_covs[k], covs[k])
+
+        assert is_spd(covs[k], decimal=7)
+
+
+@pytest.mark.parametrize(
+    "kind",
+    ["tangent", "precision", "correlation", "partial correlation"],
+)
+@pytest.mark.parametrize(
+    "cov_estimator", [EmpiricalCovariance(), LedoitWolf()]
+)
+def test_connectivity_measure_specific_for_each_kind(
+    kind, cov_estimator, signals_and_covariances
+):
+    signals, covs = signals_and_covariances
+
+    conn_measure = ConnectivityMeasure(kind=kind, cov_estimator=cov_estimator)
+    connectivities = conn_measure.fit_transform(signals)
+
+    if kind == "tangent":
+        _assert_connectivity_tangent(connectivities, conn_measure, covs)
+    elif kind == "precision":
+        _assert_connectivity_precision(connectivities, covs)
+    elif kind == "correlation":
+        _assert_connectivity_correlation(connectivities, cov_estimator, covs)
+    elif kind == "partial correlation":
+        _assert_connectivity_partial_correlation(connectivities, covs)
```

Comment:
It could be nice to put the definitions of these functions just after this test instead of at the beginning of the file. This way, the users won't have to scroll up and down. WDYT ?

Reply:
yeah I went with everything at the top rule, but if a helper function is used by only one test it would help indeed.

still don't have a good rule for organizing this.

will try to make sure things are closer to where they are used if used only once